tools/wake/build_visualization.cpp

```diff
@@ -0,0 +1,242 @@
+            os << line << " ";
+        }
+        if (!(job.outputs.empty())) {
+            os << "<br>Output file:";
+            for (const auto &output: job.outputs) {
+                os << "<br>" << output.path;
+            }
+        }
+        os << "'},\n";
+    }
+    os << "]);\n\n";
+}
+
+void write_critical_arrows(std::vector<long> &critical_path, std::ostream &os) {
+    os << "const critical_path_arrows = [\n";
+    for (size_t i = 0; i < critical_path.size() - 1; i++) {
+        os << "{ id: " << i << ", id_item_1: " << critical_path[i] << ", id_item_2: " << critical_path[i + 1]
+                  << " },\n";
+    }
+    os << "];\n\n";
+}
+
+void write_all_arrows(std::map<long, JobNode> &job_map, size_t critical_path_size, std::ostream &os) {
+    os << "const all_arrows = [\n";
+    size_t id = critical_path_size;
+    for (const auto &job_pair: job_map) {
+        JobNode job_node = job_pair.second;
+        for (long dependency: job_node.dependencies) {
+            os << "{ id: " << id << ", id_item_1: " << job_node.job.job << ", id_item_2: " << dependency
+                      << " },\n";
+            id++;
+        }
+    }
+    os << "];\n\n";
+}
+
+void write_html(std::ostream &os) {
+    std::ifstream html_template(find_execpath() + "/../share/wake/html/vis_template.html");
+    std::ifstream data(find_execpath() + "/../share/wake/html/vis_data.js");
+    std::ifstream arrow(find_execpath() + "/../share/wake/html/vis_arrow.js");
+    std::ifstream main(find_execpath() + "/../share/wake/html/vis_main.js");
+    os << html_template.rdbuf();
+    os << "<script type=\"text/javascript\">" << std::endl;
+    os << data.rdbuf();
+    os << "</script>" << std::endl;
+    os << "<script type=\"text/javascript\">" << std::endl;
+    os << arrow.rdbuf();
+    os << "</script>" << std::endl;
+    os << "<script type=\"module\">" << std::endl;
+    os << main.rdbuf();
+    os << "</script>\n"
+          "</body>\n"
+          "</html>\n";
+}
+
+void create_build_visualization(Database &db) {
+    std::vector<JobReflection> jobs = db.get_job_visualization();
+    std::map<long, JobNode> job_map;
+    for (JobReflection &job: jobs) {
+        job_map.insert({job.job, JobNode(job)});
+    }
+
+    std::vector<std::pair<int, long>> accesses = db.get_file_accesses();
+    fill_all_dependencies(accesses, job_map);
+
+    std::vector<long> top_sorted_jobs;
+    top_sort(job_map, top_sorted_jobs);
+    assign_parents(job_map, top_sorted_jobs);
+    transitive_reduction(job_map);
+
+    std::vector<long> critical_path;
+    find_critical_path(job_map, critical_path);
+
+    freopen((find_execpath() + "/../share/wake/html/vis_data.js").c_str(), "w", stdout);
+    write_jobs(job_map, std::cout);
+    write_critical_arrows(critical_path, std::cout);
+    write_all_arrows(job_map, critical_path.size(), std::cout);
+
+    freopen((find_execpath() + "/../share/wake/html/build_visualization.html").c_str(), "w", stdout);
```

Comment:
Is this the output path? It could be nice to make this configurable as we'd sometimes run this in a repository that doesn't have a `share` directory

Reply:
Will do